gnoi/gnoi.go

```diff
@@ -75,12 +82,95 @@ type otdr struct {
 
 type system struct {
 	spb.UnimplementedSystemServer
+
+	c *ygnmi.Client
+
+	rebootMu           sync.Mutex
+	hasPendingReboot   bool
+	cancelReboot       chan struct{}
+	cancelRebootFinish chan struct{}
+}
+
+func newSystem(c *ygnmi.Client) *system {
+	return &system{
+		c:                  c,
+		cancelReboot:       make(chan struct{}, 1),
+		cancelRebootFinish: make(chan struct{}),
+	}
 }
 
 func (*system) Time(context.Context, *spb.TimeRequest) (*spb.TimeResponse, error) {
 	return &spb.TimeResponse{Time: uint64(time.Now().UnixNano())}, nil
 }
 
+func (s *system) Reboot(ctx context.Context, r *spb.RebootRequest) (*spb.RebootResponse, error) {
+	if r.Method == spb.RebootMethod_POWERUP {
+		return &spb.RebootResponse{}, nil
+	}
+
+	s.rebootMu.Lock()
+	defer s.rebootMu.Unlock()
+	if s.hasPendingReboot {
```

Comment:
i think this approach is fine: however a few alternatives
1. use an atomic.Bool for hasPendingReboot simplify/eliminate the use of mutexes for these RPCs?
2. instead of having multiple chans, have a func pointer called cancelReboot, assigned in Reboot and called in CancelReboot
3. initialize the chans in the Reboot and close instead of sending empty structs

Reply:
Thanks for the suggestions. I tried each of these but didn't see a way to make them work for the reasons below. I've added comments in the code to justify using the data types chosen.

1. atomic.Bool doesn't quite work because the mutex also makes sure the sequence `check no pending reboots -> immediate reboot` is carried out atomically. If `.Load()` is called at this line, then it's possible that before the immediate reboot, that a different delayed reboot operation was scheduled such that `hasPendingReboot` is true when the immediate reboot is executed, essentially a TOCTTOU bug. Now this doesn't actually cause a practical problem in this case but semantically I think it's wrong so I prefer using the mutex.
2. I think this is an interesting idea. However the channels are needed to make CancelReboot block until the cancellation succeeds, otherwise if a reboot operation is issued too quickly after a cancel, then it might cause a conflict with the previously scheduled reboot that hasn't been fully cancelled yet.
3. I considered this, but because the spawned goroutine has a limited lifetime, I didn't see a clean way to re-create the channel for re-use after closing them. A channel cannot be closed twice.